rnd/autogpt_builder/src/components/tutorial.tsx

```diff
@@ -0,0 +1,461 @@
+import Shepherd from "shepherd.js";
+import "shepherd.js/dist/css/shepherd.css";
+
+export const startTutorial = (setPinBlocks: (value: boolean) => void) => {
+  const tour = new Shepherd.Tour({
+    useModalOverlay: true,
+    defaultStepOptions: {
+      cancelIcon: { enabled: true },
+      scrollTo: { behavior: "smooth", block: "center" },
+    },
+  });
+
+  // CSS classes for disabling and highlighting blocks
+  const disableClass = "disable-blocks";
+  const highlightClass = "highlight-block";
+  let isConnecting = false;
+
+  // Helper function to disable all blocks except the target block
+  const disableOtherBlocks = (targetBlockSelector: string) => {
+    document
+      .querySelectorAll('[data-id^="add-block-button"]')
+      .forEach((block) => {
+        block.classList.toggle(
+          disableClass,
+          !block.matches(targetBlockSelector),
+        );
+        block.classList.toggle(
+          highlightClass,
+          block.matches(targetBlockSelector),
+        );
+      });
+  };
+
+  // Helper function to enable all blocks
+  const enableAllBlocks = () => {
+    document
+      .querySelectorAll('[data-id^="add-block-button"]')
+      .forEach((block) => {
+        block.classList.remove(disableClass, highlightClass);
+      });
+  };
+
+  // Inject CSS for disabling and highlighting blocks
+  const injectStyles = () => {
+    const style = document.createElement("style");
+    style.textContent = `
+            .${disableClass} {
+                pointer-events: none;
+                opacity: 0.5;
+            }
+            .${highlightClass} {
+                background-color: #ffeb3b;
+                border: 2px solid #fbc02d;
+                transition: background-color 0.3s, border-color 0.3s;
+            }
+        `;
+    document.head.appendChild(style);
+  };
+
+  // Helper function to check if an element is present in the DOM
+  const waitForElement = (selector: string): Promise<void> => {
+    return new Promise((resolve) => {
+      const checkElement = () => {
+        if (document.querySelector(selector)) {
+          resolve();
+        } else {
+          setTimeout(checkElement, 10);
```

Comment:
What’s this doing here

Reply:
Why are setting timeout? Just slows stuff down. Check the state, trigger based on it if needed

Reply:
While I agree that it'd be better to do it in a reactive way, I don't think this is a big deal, it's 10ms and only during the tutorial, so if it takes a lot of time to reimplement, this can stay as is imo.

Reply:
because the ui constantly updates we have this to check for the addition of items during the steps, its only used during the tutorial, i can look into a better setup but this current one works well